### PR TITLE
Move and refactor clock components

### PR DESCRIPTION
**Pull request by tidal**
- Moves clock components from `SDK` into `SDK\Common\Time` namespace
- Adds BC compatible class alias' clock components
- Introduces and uses `StopWatch` class
- Adds `Factory` for Clock and other time related components
- Refactors and optimizes `SystemClock` (wall-time is now calcualted with nanosecond precision)
- Moves time conversion methods into `Util` class
- Deprecates `AbstractClock`

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
     "config": {
         "sort-packages": true,
         "allow-plugins": {
-            "composer/package-versions-deprecated": true
+            "composer/package-versions-deprecated": true,
+            "sllh/composer-versions-check": false
```
Comment (tidal) [inline composer.json:26]: This should not be there. I will remove it.

File: tests/Unit/SDK/Common/Time/ClockFactoryTest.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        $this->assertInstanceOf(SystemClock::class, ClockFactory::getDefault());
+    }
+
+    public function test_default_is_settable(): void
+    {
+        $clock = $this->createMock(ClockInterface::class);
+        ClockFactory::setDefault($clock);
+
+        $this->assertEquals($clock, ClockFactory::getDefault());
```
Comment (brettmc) [inline tests/Unit/SDK/Common/Time/ClockFactoryTest.php:32]: I think this should be `assertSame` to confirm they are the same instance

File: tests/Unit/SDK/Common/Time/ClockFactoryTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public function test_default_is_resettable(): void
+    {
+        ClockFactory::setDefault(
+            $this->createMock(ClockInterface::class)
+        );
+        ClockFactory::setDefault(null);
+
+        $this->assertInstanceOf(SystemClock::class, ClockFactory::getDefault());
```
Comment (brettmc) [inline tests/Unit/SDK/Common/Time/ClockFactoryTest.php:42]: and an assertNotSame on the original mock?

**Review by brettmc — APPROVED** [review]

File: tests/Unit/SDK/Common/Time/ClockFactoryTest.php (hunk 3 above)
Comment (tidal) [reply tests/Unit/SDK/Common/Time/ClockFactoryTest.php:42]: Why? This is already covered in above test, or what do you mean?

File: tests/Unit/SDK/Common/Time/StopWatchFactoryTest.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        $this->assertInstanceOf(StopWatch::class, StopWatchFactory::getDefault());
+    }
+
+    public function test_default_is_settable(): void
+    {
+        $stopwatch = $this->createMock(StopWatchInterface::class);
+        StopWatchFactory::setDefault($stopwatch);
+
+        $this->assertEquals($stopwatch, StopWatchFactory::getDefault());
```
Comment (brettmc) [inline tests/Unit/SDK/Common/Time/StopWatchFactoryTest.php:37]: assertSame

File: tests/Unit/SDK/Common/Time/ClockFactoryTest.php (hunk 3 above)
Comment (brettmc) [reply tests/Unit/SDK/Common/Time/ClockFactoryTest.php:42]: I mean confirm that setDefault(null) did something by asserting that getDefault() has returned something _different_ to the original default set

Comment (tidal) [reply tests/Unit/SDK/Common/Time/ClockFactoryTest.php:42]: Ah, ok got ya!

File: tests/Unit/SDK/Common/Time/ClockFactoryTest.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    public function test_default_is_resettable(): void
+    {
+        $clock = $this->createMock(ClockInterface::class);
+        ClockFactory::setDefault(
+            $clock
+        );
+        ClockFactory::setDefault(null);
+
+        $this->assertNotSame($clock, ClockFactory::getDefault());
```
Comment (tidal) [inline tests/Unit/SDK/Common/Time/ClockFactoryTest.php:43]: Like this?

Comment (brettmc) [reply tests/Unit/SDK/Common/Time/ClockFactoryTest.php:43]: yes!

**Comment by tidal** [comment]
Thanks for the review @brettmc !!!

**Pull request merged**